rust/common-graph/src/graph.rs

```diff
@@ -0,0 +1,533 @@
+                PortDetails {
+                    node: self.label().into(),
+                    port: port_name.into(),
+                    port_type: PortType::Input,
+                }
+                .into(),
+            )),
+            // Port exists and has no outgoing connection
+            None => {
+                port.1 = Some(port_ref);
+                Ok(())
+            }
+        }
+    }
+}
+
+/// Big TODO
+#[derive(Debug, Default)]
+pub struct PortGraph<T: Debug> {
+    nodes: Vec<Node<T>>,
+}
+
+impl<T> PortGraph<T>
+where
+    T: Debug,
+{
+    /// Takes an iterator of tuples of nodes containing an
+    /// inner `T`, a label used in debugging, and its input
+    /// and output port names. The iterator of edges is used to
+    /// populate all connections. Fails if graph is invalid.
+    pub fn new<NodeIter, EdgeIter, S>(nodes_iter: NodeIter, edges: EdgeIter) -> Result<Self>
+    where
+        NodeIter: IntoIterator<Item = (Option<T>, S, (Vec<String>, Vec<String>))>,
+        EdgeIter: IntoIterator<Item = Edge>,
+        S: Into<String>,
+    {
+        let mut nodes = vec![];
+        let mut node_set: HashSet<String> = HashSet::default();
+        for (index, (inner, label, io)) in nodes_iter.into_iter().enumerate() {
+            let label = label.into();
+            if !node_set.insert(label.clone()) {
+                return Err(CommonGraphError::DuplicateLabels(label.clone()));
+            }
+            let node = Node::new(index, inner, label, io.0, io.1)?;
+            if index == 0 && !node.is_root() {
+                return Err(CommonGraphError::InvalidRoot);
+            }
+            if index != 0 && node.is_root() {
+                return Err(CommonGraphError::InvalidRoot);
+            }
+            nodes.push(node);
+        }
+
+        // A graph that doesn't have both a root node and one other node
+        // is considered empty.
+        if nodes.len() < 2 {
+            return Err(CommonGraphError::EmptyGraph);
+        }
+
+        populate_connections(&mut nodes, edges)?;
+
+        for node in nodes.iter() {
+            if is_empty(node.inputs()) && is_empty(node.outputs()) {
+                return Err(CommonGraphError::OrphanedNode(node.label().into()));
+            }
+        }
+
+        Ok(PortGraph { nodes })
+    }
+
+    /// Returns a reference to nodes in this graph.
+    pub fn nodes(&self) -> &Vec<Node<T>> {
+        &self.nodes
+    }
+
+    /// Returns a reference to an internal [Node].
+    pub fn get_node(&self, index: usize) -> Result<&Node<T>> {
+        self.nodes
+            .get(index)
+            .ok_or_else(|| CommonGraphError::MissingNode(index))
+    }
+
+    /// Returns a [GraphProcessor] that can be iterated over
+    /// to propagate `input` throughout the graph network.
+    pub fn process<'a, V, I>(&'a self, input: I) -> Result<GraphProcessor<'a, T, V>>
+    where
+        I: IntoIterator<Item = (&'a str, V)>,
+        V: Clone + 'a,
```

Comment:
I don't like the idea of `V` needing to be cloneable; could probably work around this